rust-runtime/aws-smithy-observability/src/attributes.rs

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Attributes (also referred to as tags or annotations in other telemetry systems) are structured
+//! key-value pairs that annotate a span or event. Structured data allows observability backends
+//! to index and process telemetry data in ways that simple log messages lack.
+
+use std::collections::HashMap;
+
+/// The valid types of values accepted by [Attributes].
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq)]
+pub enum AttributeValue {
+    /// Holds an [i64]
+    I64(i64),
+    /// Holds an [f64]
+    F64(f64),
+    /// Holds a [String]
+    String(String),
+    /// Holds a [bool]
+    Bool(bool),
+}
+
+/// Structured telemetry metadata.
+#[non_exhaustive]
+#[derive(Clone)]
+pub struct Attributes {
+    attrs: HashMap<String, AttributeValue>,
+}
+
+impl Attributes {
+    /// Create a new empty instance of [Attributes].
+    pub fn new() -> Self {
+        Self {
+            attrs: HashMap::new(),
+        }
+    }
+
+    /// Set an attribute.
+    pub fn set(&mut self, key: String, value: AttributeValue) {
```

Comment:
Should probably take more flexible types like `impl Into<String>` or `impl Into<AttributeValue>`

Reply:
Done in https://github.com/smithy-lang/smithy-rs/pull/3912/commits/0559e8b54fb41bcf04125e2e3cedeec87b24a9f9